<a href="https://colab.research.google.com/github/Alejom74/Business-Inteligence/blob/main/Reto_1_Alejandro_Mariaca.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Se procede a cargar las librerias de trabajo

In [ ]:
import numpy as np
import pandas as pd

1. Carga la base de datos

In [ ]:
nxl='/content/1. BD1_Longitudinal Survey NB.xlsx'
XDB=pd.read_excel(nxl,sheet_name=1)
XDB.head(100)

# Se seleccionan las variables de interes
XDB=XDB[['Age','Urban','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']]
XDB.head(10)

# Eliminar filas con al menos un valor faltante
XDB=XDB.dropna()

# Se separan las variables de entrada y salida
XD=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']] #Variables de Entrada
yd=XDB[['Urban']] #Variable de Salida

#RESPUESTA PRIMERA PREGUNTA
total_datos = XDB.shape[0]
print("El número total de datos en la base de datos actualizada es:", total_datos)

#RESPUESTA SEGUNDA PREGUNTA
# Porcentaje de cada categoría en Urban
categories = XDB['Urban'].value_counts(normalize=True) * 100
print('Porcentaje de datos en cada categoría:\n', categories)

#RESPUESTA SIGUIENTES PREGUNTAS
# Intervalos de variación de Edad, Altura y Peso de las personas
print('Rango de Edad', (XDB['Age'].min(), XDB['Age'].max()))
print('Rango de Altura', (XDB['Height'].min(), XDB['Height'].max()))
print('Rango de Peso', (XDB['Weight'].min(), XDB['Weight'].max()))


El número total de datos en la base de datos actualizada es: 5948
Porcentaje de datos en cada categoría:
 Urban
1.0    79.169469
0.0    20.830531
Name: proportion, dtype: float64
Rango de Edad (15, 22)
Rango de Altura (51.0, 80.0)
Rango de Peso (72.0, 375.0)


2. Se procede con la implementacion del modelo

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB() #Aloja el Modelo
mnb.fit(XD,yd) #Entrena el Modelo

#Obtenemos la informacion del Modelo
u=mnb.theta_ #Obtenemos la media de las variables socioeconomicas
print("La media de los ingresos es :\n",u)
var=mnb.var_ #Obtenemos la varianza de las variables socioeconomicas
print("La varianza de los ingresos es :\n",var)
sigma=np.sqrt(var) #Obtenemos la desviacion estandar de las variables socioeconomicas
print("La desviacion estandar de los ingresos es :\n",sigma)

#Se determinan los limites superiores e inferiores
LS=u+sigma
print("El limite superior es :\n",LS)
LI=u-sigma
print("El limite inferior es :\n",LI)

#Evaluamos EL SOLICITANTE REQUERIDO
XDS=[[34,12,15,2,24,69,160,48000]]
ydp=mnb.predict(XDS) #Se obtiene la prediccion del modelo
print("La prediccion del modelo es :\n",mnb.predict(XDS))

if ydp==0:
  print("La persona vive en Area Rural")

if ydp==1:
  print("La persona vive en Area Urbana")

La media de los ingresos es :
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
La varianza de los ingresos es :
 [[6.11377773e+00 9.96419474e+00 1.51194652e+01 6.51406931e+00
  1.77789403e+01 1.66650620e+01 9.21039647e+02 1.00161093e+09]
 [6.11792319e+00 1.15645045e+01 1.69373304e+01 5.96364977e+00
  1.76201044e+01 1.76562243e+01 9.01712618e+02 1.30246862e+09]]
La desviacion estandar de los ingresos es :
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
El limite superior es :
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.1

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


3. Evaluamos el desempeño del modelo

In [ ]:
from sklearn.metrics import confusion_matrix

#Haga un barrido de los datos
vdp2=mnb.predict(XD) #Haga el pronóstico para los datos base
cm=confusion_matrix(yd,vdp2) #Cree la matriz de confusión
print("La matriz de confusión es:\n",cm)
VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]

#Metricas Desempeño
Ex=(VP+VN)/(VP+VN+FP+FN) #Exactitud
print("La exactitud del modelo es:\n",Ex)

TE=(FN+FP)/(VP+VN+FP+FN) #Tasa de Error
print("El error del modelo es:\n",TE)

Se=VP/(VP+FN) #Sensibilidad
              #Sobre el total de los urbanos
print("La sensibilidad del modelo es:\n",Se)

Sp=VN/(VN+FP) #Especificidad
              #Sobre el total de los no urbanos (rurales)
print("La especificidad del modelo es:\n",Sp)

Ur=VP/(VP+FP) #Precisión
             #Sobre el total del FP + VP pronóstico
print("La precisión del modelo es:\n",Ur)

Ru=VN/(VN+FN) #Precisión Negativa
               #Sobre el total del FN + VN pronóstico
print("La precisión negativa del modelo es:\n",Ru)

La matriz de confusión es:
 [[   3 1236]
 [   8 4701]]
La exactitud del modelo es:
 0.7908540685944856
El error del modelo es:
 0.20914593140551446
La sensibilidad del modelo es:
 0.9983011255043533
La especificidad del modelo es:
 0.002421307506053269
La precisión del modelo es:
 0.7918140474987367
La precisión negativa del modelo es:
 0.2727272727272727


**Analisis de Resultados:**
De acuerdo con la base de datos suministrada se tomaron como referencia para identificar si las personas viven en el Area Urbana o no (Rural) las variables socioeconómicas: Edad, Educación de los Padres, Estatura, Peso, Ingresos, Escala de Autoestima, Tamaño de la Familia. La base de datos posee un total de 5948 datos, el cual de este numero, el 79.16% de las personas viven en Áreas Urbanas y el porcentaje restante (20.83%) viven en Areas Rurales.  De manera general, el modelo alcanzo una Exactitud cercana al 80% en promedio (79.1%), lo que muestra un comportamiento bastante aceptable frente a la informacion de las personas que viven en Areas Urbanas. Tambien es importante mencionar que el modelo tiene una Especificidad del 0%, lo que muestra su pesimo comportamiento frente a las personas que viven en Áreas Rurales, implicaría que el modelo está clasificando erróneamente a muchas personas rurales como urbanas. Pero algo a resaltar que es muy positivo es el tema de la Sensibilidad (99%), en este caso, el valor alto de sensibilidad indica que el modelo es excelente para identificar correctamente a las personas que viven en área urbana. Este buen comportamiento fue promovido por el indice de precision el cual alcanzo un valor cercano al 80% en promedio (79%), lo que muestra el buen criterio del modelo frente a la información brindada con las personas que viven en Area Urbana.